src/libstd/io/stdio.rs

```diff
@@ -155,6 +143,121 @@ fn handle_ebadf<T>(r: io::Result<T>, default: T) -> io::Result<T> {
     }
 }
 
+enum StdioBufferKind {
+    LineBuffered,
+    Buffered,
+    Unbuffered,
+}
+
+/// A reader that can either be buffered or fake.
+///
+/// If it is fake, all reads succeed and indicate end-of-file.
+enum StdioReader<R: Read> {
+    Buffered(BufReader<HandleEbadf<R>>),
+    Fake,
+}
+
+impl<R: Read> StdioReader<R> {
+    fn new_buffered(reader: R) -> StdioReader<R> {
+        let cap = stdio::STDIN_BUF_SIZE;
+        StdioReader::Buffered(BufReader::with_capacity(cap, HandleEbadf(reader)))
+    }
+    fn new_fake() -> StdioReader<R> {
+        StdioReader::Fake
+    }
+}
+
+impl<R: Read> Read for StdioReader<R> {
+    fn read(&mut self, buf: &mut [u8]) -> io::Result<usize> {
+        match self {
+            StdioReader::Buffered(i) => i.read(buf),
+            StdioReader::Fake => Ok(0),
+        }
+    }
+    fn read_vectored(&mut self, bufs: &mut [io::IoSliceMut<'_>]) -> io::Result<usize> {
+        match self {
+            StdioReader::Buffered(i) => i.read_vectored(bufs),
+            StdioReader::Fake => Ok(0),
+        }
+    }
+    #[inline]
+    unsafe fn initializer(&self) -> Initializer {
+        match self {
+            StdioReader::Buffered(i) => i.initializer(),
+            StdioReader::Fake => Initializer::nop(),
+        }
+    }
+}
+
+impl<R: Read> BufRead for StdioReader<R> {
+    fn fill_buf(&mut self) -> io::Result<&[u8]> {
+        match self {
+            StdioReader::Buffered(i) => i.fill_buf(),
+            StdioReader::Fake => Ok(&[]),
+        }
+    }
+    fn consume(&mut self, amt: usize) {
+        match self {
+            StdioReader::Buffered(i) => i.consume(amt),
+            StdioReader::Fake => (),
+        }
+    }
+}
+
+/// A writer that can either be line-buffered, buffered, unbuffered or fake.
+///
+/// If it is fake, all outputs just succeed and are dropped silently.
+enum StdioWriter<W: Write> {
```

Comment:
There was a comment in the original `FIXME` suggesting we should call `flush` when dropping the handle if we're panicking. Did you have any thoughts on this @tbu-?